test: settings: functional: fix missing test_main

yperess opened the pull request:
The functional tests for settings subsystem regressed with the
introduction of the weak test_main function. This was caused by
Issue #19582 since the common code was being linked as a library.

Update the logic of the common code to directly add the sources
to the `app` target. This resolved both Issue #19582 and #42396

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/42396
Fixes #19582

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
test: settings: functional: fix missing test_main

The functional tests for settings subsystem regressed with the
introduction of the weak test_main function. This was caused by
Issue #19582 since the common code was being linked as a library.

Update the logic of the common code to directly add the sources
to the `app` target. This resolved both Issue #19582 and #42396

Signed-off-by: Yuval Peress <peress@google.com>

nvlsianpu commented:
@yperess 
Hi,
I think more expected is to provide a patch which will allow main_test() function in tests-suite's own library. Sorry that I didn't describe that expectation in the issue.

yperess commented:
> @yperess Hi, I think more expected is to provide a patch which will allow main_test() function in tests-suite's own library. Sorry that I didn't describe that expectation in the issue.

I see what you're saying, but I don't think that's possible. The linking order matters here. One alternative (which I don't mind changing this PR to) is to inverse the relationship. Basically add `settings_basic_test.c` as the source for `app` and have the various tests (`fcb`, `file`, and `nvs`) create the library which is linked into `app`

yperess commented:
https://stackoverflow.com/questions/51656838/attribute-weak-and-static-libraries/51970423#51970423 actually gives a very good description of what's going on under the hood.

yperess commented:
Ping

gopiotr reviewed (changes requested):
Your commit fix only those two tests:
```
tests/subsys/settings/functional/nvs
tests/subsys/settings/functional/fcb
```
During works connected with this PR: https://github.com/zephyrproject-rtos/zephyr/pull/42482 I discovered, that exactly the same problems occur in following tests:
```
tests/subsys/settings/fcb/raw
tests/subsys/settings/littlefs/base64
tests/subsys/settings/littlefs/raw
tests/subsys/settings/nvs/raw
```
Please consider of fixing those bugs in all those problematic tests from `tests/subsys/settings/` directory in this PR

Commit:
tests: settings: Remove invalid config

Tests running on native_posix and native_posix_64 are selecting
CONFIG_STDOUT_CONSOLE, but this config value depends on
!NATIVE_APPLICATION. This configuration causes the CI to fail.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
This PR now fixes all the tests under `tests/subsys/settings/`

tejlmand commented on tests/subsys/settings/fcb/raw/prj_native_posix.conf:
```diff
@@ -1,5 +1,4 @@
 CONFIG_ZTEST=y
-CONFIG_STDOUT_CONSOLE=y
```
Question, did you investigate whether [testcase.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/3631fc4ed87d93cceb4affeb2843ff91897ce430/tests/subsys/settings/fcb/raw/testcase.yaml) should be updated and split in two parts.

One for the native posix platforms, and one for other platforms where you can then add:
```
extra_configs:
- CONFIG_STDOUT_CONSOLE=y
```

I assume `CONFIG_STDOUT_CONSOLE=y` was added originally for a reason.

yperess replied:
I did not, seemed unrelated. I went ahead and added that as a separate commit on this PR to clean it up.

Commit:
tests: settings: Update test project structures

Update the structure to reduce redundancies with additional config
files needed for native_posix platforms that replicate most of the
same config values.

Signed-off-by: Yuval Peress <peress@google.com>

tejlmand reviewed (approved):
Looks good, Thanks for the last cleanup commit :+1:

yperess commented:
> Looks good, Thanks for the last cleanup commit +1

Accidentally left in the `native_posix` platforms in `littlefs/raw`. Fixed it in the latest patch

nvlsianpu reviewed (commented):
Checked that it actually works well.
post merge. +1
